### DISCUSSION TITLE
Replace webpack by vite in the front part

### DISCUSSION THREAD

**Discussion by csouchet** [apache/superset#25711] [Ideas]
Would it be possible to replace `webpack` with a faster build tool, such as [Vite](https://vitejs.dev/)?

I've tried running the live-reload mode of the Superset front-end in version `3.0.1` (without any modifications), running on Node.js `v16.20.2` (EOL). 

`webpack` makes the build very time-consuming in both production and development modes. 
The advantage of vite is that it only rebuilds what has been changed, making reloading instantaneous in development mode.

In addition, there are no problems with the build process. However, I am getting connection errors when trying to load the page, the cause of which is unclear to me:
```shell
[HPM] Proxy created: / -> http://localhost:8088
[HPM] Error occurred while proxying request localhost:9000/ to http://localhost:8088/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
[HPM] Error occurred while proxying request localhost:9000/favicon.ico to http://localhost:8088/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
```

**Comment by NurramoX** [comment]
I'm also interested in the potential shift from webpack to Vite. It's an appealing prospect, especially for a project of Superset's scale where webpack's build times are becoming increasingly cumbersome. Vite's behind-the-scenes use of Rollup and ESBuild, known for their speed and efficiency, could potentially improve the development workflow a lot. The industry also seems to be gradually pivoting away from webpack in favor of more performant alternatives like Vite.

Addressing your issue @csouchet, could you confirm if the backend server was actively up and running on port 8088 at the time of encountering these errors? Additionally, I think it would be highly insightful if you could walk us through the exact steps you took leading up to this error.

**Comment by harksin** [comment]
Hello Interested in this shift too.

Which kind of validation should we ask for before openning a PR for this ?

**Comment by manatik** [comment]
I managed to raise the superset with vite. It works in dev mode. The build is successful after several fixes of imports in less files. But it refuses to work in the final image. Maybe someone can fix it. Below I attach the vite configuration

library versions
```json
"vite": "^5.4.14",
"vite-tsconfig-paths": "^5.1.4",
"@vitejs/plugin-react": "^4.3.4",
"rollup-plugin-visualizer": "^5.14.0",
```

```ts
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import tsconfigPaths from 'vite-tsconfig-paths';
import { visualizer } from 'rollup-plugin-visualizer';
import { resolve } from 'path';

const isDevMode = process.env.NODE_ENV !== 'production';
const ASSET_BASE_URL = process.env.ASSET_BASE_URL || '';

export default defineConfig({
  base: `${ASSET_BASE_URL}/static/assets/`,
  plugins: [
    react(),
    tsconfigPaths(),
    process.env.BUNDLE_ANALYZER && visualizer({ open: true }),
  ],
  server: {
    port: 9000,
    proxy: {
      '/': {
        target: process.env.SUPERSET_URL || 'http://localhost:8088',
        changeOrigin: true,
        selfHandleResponse: false,
        configure: (proxy) => {
          proxy.on('proxyRes', (proxyRes, req, res) => {
            Object.keys(proxyRes.headers).forEach((key) => {
              if (key.toLowerCase() !== 'content-length') {
                res.setHeader(key, proxyRes.headers[key]);
              }
            });

            if (proxyRes.headers['set-cookie']) {
              const newCookies = proxyRes.headers['set-cookie'].map((cookie) =>
                cookie.replace(/Domain=[^;]+;?/i, '')
              );
              res.setHeader('set-cookie', newCookies);
            }
          });
        },
      },
    },
  },
  build: {
    outDir: '../superset/static/assets',
    rollupOptions: {
      input: {
        preamble: resolve(__dirname, 'src/preamble.ts'),
        theme: resolve(__dirname, 'src/theme.ts'),
        menu: resolve(__dirname, 'src/views/menu.tsx'),
        spa: resolve(__dirname, 'src/views/index.tsx'),
        embedded: resolve(__dirname, 'src/embedded/index.tsx'),
      },
      output: {
        entryFileNames: '[name].[hash].js',
        chunkFileNames: '[name].[hash].js',
        manualChunks(id) {
          if (id.includes('node_modules')) {
            return 'vendors';
          }
          if (id.includes('thumbnail') && (id.endsWith('.png') || id.endsWith('.jpg'))) {
            return 'thumbnail';
          }
          return null;
        },
      },
    },
  },
  resolve: {
    alias: {
      react: resolve(__dirname, 'node_modules/react'),
      'react-dom': resolve(__dirname, 'node_modules/react-dom'),
      antd: resolve(__dirname, 'node_modules/antd'),
    },
  },
  css: {
    preprocessorOptions: {
      less: {
        javascriptEnabled: true,
        paths: ['node_modules']
      },
    },
  },
});
```

**Comment by hainenber** [comment]
I don't have any strong opinions on build tooling decision but any migration shouldn't disrupting usual workflow and be feature parity with existing webpack. I've done some experimentations with various builders like vite, rspack and biome and realize greenfield projects would be most beneficial to them and not brownfield ones as there are way too many historical configurations that I'm sure not all of us maintainers can recall.
